### ISSUE TITLE
synth.getMidiFile generates "Paused before potential out-of-memory crash" in Chrome during call to warn

### ISSUE THREAD

**Issue by johnkolen** [paulrosen/abcjs#689]
While using abcjs basic 6.0.0.beta.33, the following initialization
```
var abcOptions = {
  add_classes: true,
  clickListener: self.clickListener,
  responsive: "resize"
};
var visualObj = ABCJS.renderAbc("sheetmusic",
 			                                 $("#sheetmusic").html(),
				                         abcOptions)[0];
var midi = ABCJS.synth.getMidiFile($("#sheetmusic").html());
```
causes Chrome Version 91.0.4472.77 (Official Build) (x86_64) on iOS to generate a "Paused before potential out-of-memory crash" at
```
// line 2685
parseCommon.gsub = function (source, pattern, replacement) {
  return source.split(pattern).join(replacement);
};
```
Source is a 932kb string, pattern = "<", and replacement = "&lt"

Call Stack:
```
parseCommon.gsub | @ | abcjs_basic_6.0.0-beta.33.js:2686
encode | @ | abcjs_basic_6.0.0-beta.33.js:2961
warn | @ | abcjs_basic_6.0.0-beta.33.js:2968
MusicParser.parseMusic | @ | abcjs_basic_6.0.0-beta.33.js:7527
parseLine | @ | abcjs_basic_6.0.0-beta.33.js:3255
Parse.parse | @ | abcjs_basic_6.0.0-beta.33.js:3384
tunebook.renderEngine | @ | abcjs_basic_6.0.0-beta.33.js:690
getMidiFile | @ | abcjs_basic_6.0.0-beta.33.js:15288
```

<br class="Apple-interchange-newline">
Rending the sheetmusic is fine, the browser has no complaints. When trying to attach a midi player it has problems. It looks like it finds a character it doesn't like and then wants to generate a warning, but the source is way to big.

Here's the ABC that is being processed 
```
X:1
T:Faceted Word Class
M:4/4
Q:1/4=140
V:melody clef=treble name=melody
V:harmony clef=treble name=harmony
K:GMix
[V:melody]
d'2f'2b2d'2|a2g'2g2d'2|b2b2e'2c'2|c2f'2a2g'2|
[V:harmony]
B2d2f2g2|f2e2c2A2|A2F2D2B,2|B,2D2E2G2|
[V:melody]
e2a2c'2d'2|d'2G2d2G2|d2G2B2G2|A2B2e2A2|
[V:harmony]
[A,CEG]8|G,2A,2C2E2|E2C2B,2G,2|B,2D2F2G2|
[V:melody]
e2c2e'2f2|c2f2c4|G2BG3d2|GAdGF2a
[V:harmony]
A2F2D2C2|[G,B,DF]8|G,2B,2D2F2|C2A,2F,2E,2|
[V:melody]
c2gdfceA|AA2ec3e-|egeaec2E|B,g2BG3D-|D
[V:harmony]
B,2D2F2G2|F2E2C2A,2|[B,,D,F,A,]8|E,2G,2A,2C2|
[V:melody]
AEB,GD2GD-|DGAdAecE|A,2ef3c'2|fcfcG,4|
[V:harmony]
A,2F,2E,2C,2|B,,2D,2E,2G,2|F,2D,2C,2A,,2|[G,,B,,D,F,]8|
[V:melody]
DG,d2GdGc|d2G3BG2|BGdGcdgc-|cfG2AdAE|
[V:harmony]
G,,2A,,2C,2E,2|G,2B,2C2E2|F2D2B,2G,2|F,2A,2C2D2|
[V:melody]
c2G3BG2|DG,d2GdGc|d2G3BG2|cGdGcdg
[V:harmony]
[E,G,B,D]8|G,,2A,,2C,2E,2|G,2B,2C2E2|[G,B,DF]8|
[V:melody]
c2fG2AdA|EG7-|Gd'/2g/2b/2g/2d/2b/2g/2b/2g/2d/2g/2g/2d/2g/2|d/2g/2a/2d/2G/2d/2G/2B/2G/2f/2dcz/2g/2-|g/2a/2
[V:harmony]
D2C2A,2F,2|[G,B,DF]8|D2F2G2B2|F2D2B,2A,2|
[V:melody]
e/2B/2d/2G/2A/2d/2G/2b/2e/2a/2e/2g/2d/2a/2f/2c'/2|a/2z/2e/2a/2zf/2e/2Az3-|z/2a/2A/2c'/2e/2c/2f/2e/2z/2F/2z3|z3G/2z/2d/2g/2d/2g/2d/2G,/2d/2
[V:harmony]
G,2B,2C2E2|F2E2C2A,2|[G,B,DF]8|[G,,B,,D,F,]8|
[V:melody]
z7/2Gd/2G/2zB/2G/2z/2|d/2G/2d/2e/2A/2c/2G/2d/2G/2a/2E/2c/2E/2c/2E/2a/2|e/2a/2ez/2AeA/2E/2A/2F/2c/2F/2C/2|f/2c/2E/2f/2c/2e/2f/2c'/2g/2e/2a/2f/2c/2e/2f/2a/2-|a/2
[V:harmony]
G,2A,2C2E2|C2B,2G,2E,2|C,2D,2F,2A,2|[G,,B,,D,F,]8|
[V:melody]
d'/2g/2b/2g/2d/2b/2g/2b/2g/2d/2g/2g/2d/2g/2d/2g/2|a/2d/2G/2d/2G/2B/2G/2f/2dcz/2ga/2|e/2B/2d/2G/2A/2d/2G/2b/2e/2a/2e/2g/2d/2a/2f/2c'/2|a/2z/2e/2a/2zf/2e/2A
[V:harmony]
D2F2G2B2|F2D2B,2A,2|G,2B,2C2E2|F2E2C2A,2|
[V:melody]
z7/2a/2A/2c'/2e/2c/2f/2e/2z/2F/2|z3z3G/2z/2d/2g/2|d/2g/2d/2G,/2d/2z7/2Gd/2G/2|zd/2G/2z/2d/2G/2D/2g/2A,/2f/2c/2g/2D/2e'/2A/2|c/2E/2c/2E/2F/2c'/2
[V:harmony]
[G,B,DF]8|G,,2B,,2D,2F,2|G,2A,2C2E2|G,2E,2C,2A,,2|
[V:melody]
f/2cz/2Efc/2e/2f/2A/2e/2f/2A/2c'/2|d'8|G2BG3d2|GAdGF2a
[V:harmony]
F,2A,2C2D2|[GBdf]8|G,2B,2D2F2|C2A,2F,2E,2|
[V:melody]
c2gdfceA|AA2ec3e-|egeaec2E|B,g2BG3D-|D
[V:harmony]
B,2D2F2G2|F2E2C2A,2|[B,,D,F,A,]8|E,2G,2A,2C2|
[V:melody]
AEB,GD2GD-|DGAdAecE|A,2ef3c'2|fcfcG,4|
[V:harmony]
A,2F,2E,2C,2|B,,2D,2E,2G,2|F,2D,2C,2A,,2|[G,,B,,D,F,]8|
[V:melody]
DG,d2GdGc|d2G3BG2|BGdGcdgc-|cfG2AdAE|
[V:harmony]
G,,2A,,2C,2E,2|G,2B,2C2E2|F2D2B,2G,2|F,2A,2C2D2|
[V:melody]
c2G3BG2|DG,d2GdGc|d2G3BG2|cGdGcdg
[V:harmony]
[E,G,B,D]8|G,,2A,,2C,2E,2|G,2B,2C2E2|[G,B,DF]8|
[V:melody]
c2fG2AdA|EG7-|Gd'2f'2b2d'-|d'a2g'2g2d'-|d'
[V:harmony]
D2C2A,2F,2|[G,B,DF]8|B2d2f2g2|f2e2c2A2|
[V:melody]
b2b2e'2c'2|c2f'2a2g'2|e2a2c'2d'2|d'2G2g2d2|
[V:harmony]
A2F2D2B,2|B,2D2E2G2|[A,CEG]8|G,2A,2C2E2|
[V:melody]
g2d2g2B2|B2B2e2c2|f'2a2e'2a2|e2f2a2c2|
[V:harmony]
G2E2D2B,2|B,2C2E2G2|E2G2A2c2|[CEGB]8|
[V:melody]
c'2g2d2G2|d2G2D2g2|B,2G,2c2D2|g2a2d'2a2|
[V:harmony]
[G,B,DF]8|C,2E,2G,2A,2|E,2C,2B,,2G,,2|D2F2A2c2|
[V:melody]
d8|
[V:harmony]
[G,B,DF]8|
```

**Comment by johnkolen** [comment]
After poking about, it looks like warn() doesn't work well with very long lines. Limiting the before and after regions made the OOM warning go away.
```
  var warn = function warn(str, line, col_num) {
   ...
    var clean_line = encode(line.substring(col_num - 64, col_num)) + '<span style="text-decoration:underline;font-size:1.3em;font-weight:bold;">' + bad_char + '</span>' + encode(line.substring(col_num + 1).substring(0,64));
```
Also, how do you specify where the warnings get dumped? It looks liken you can set warnings_id for Editor, but I could not find a way for display & midi.

**Comment by paulrosen** [comment]
I made that change and it will be in the next beta. Thanks!

For getting the warnings, they are in the tune object, so look in `visualObj.warnings`.

**Comment by paulrosen** [comment]
Fixed in version 6.0.0-beta.35
